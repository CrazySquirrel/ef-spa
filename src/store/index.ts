declare const window: any;
declare const global: any;

import {createStore, compose} from 'redux';
import {Action} from 'redux-actions';

import {RouteHandle, RouteStore} from 'store/reducers/router/index';
import {AsideHandle, AsideStore} from 'store/reducers/aside/index';
import {CSRFHandle, CSRFStore} from 'store/reducers/csrf/index';

// istanbul ignore next
export const localWindow: any = typeof window === 'undefined' ? (typeof global === 'undefined' ? global : {}) : window;

export interface StoreTree extends RouteStore,
    AsideStore,
    CSRFStore {
  modified: string;
}

export const Handlers = [
  RouteHandle,
  AsideHandle,
  CSRFHandle,
];

export const composeEnhancers = localWindow.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

export const DefaultStore = {
  modified: '',

  location: '/',

  csrf: '',
};

export const Store = createStore(
    (state: StoreTree, action: Action<{}>): StoreTree => {
      return Handlers.reduce(
          (newState: StoreTree, handler: (...props: any[]) => any) => handler(newState, action),
          state,
      ) as StoreTree;
    },
    localWindow.__PRELOADED_STATE__ || DefaultStore,
    composeEnhancers(),
);
