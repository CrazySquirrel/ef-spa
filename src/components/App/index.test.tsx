import * as React from 'react';

import {shallow, mount} from 'enzyme';

import {Provider} from 'react-redux';
import {BrowserRouter} from 'react-router-dom';

import {StoreTree, DefaultStore} from '../../store';

import {
  MetadataImage,
  MetadataVideo,
  MetadataAudio,
  MetadataType,
} from '../../types';

import {mapStateToProps, mapDispatchToProps, App} from './index';

import configureStore from 'redux-mock-store';

const mockStore = configureStore();

const data = 'Sat Jan 12 2019 02:56:35 GMT+0300 (Moscow Standard Time)';

describe('App', () => {
  const defaultProps = {
    title: 'test',
    description: 'test',
    keywords: 'test',
    location: '/',
    modified: data,
  };

  it('render default', () => {
    expect(shallow(<App {...defaultProps} />)).toMatchSnapshot();
  });

  it('render article', () => {
    expect(shallow(<App {...defaultProps} type={MetadataType.ARTICLE}/>)).toMatchSnapshot();
  });

  it('render with update', () => {
    const additionalProps = {
      update: (location: string) => location,
    };

    expect(shallow(<App {...defaultProps} {...additionalProps}/>)).toMatchSnapshot();
  });

  it('render with audios', () => {
    const audios: MetadataAudio[] = [{
      src: 'test',
      type: 'test',
    }];

    expect(shallow(<App {...defaultProps} audios={audios}/>)).toMatchSnapshot();
  });

  it('render with videos', () => {
    const videos: MetadataVideo[] = [{
      src: 'test',
      type: 'test',
      width: 'test',
      height: 'test',
    }];

    expect(shallow(<App {...defaultProps} videos={videos}/>)).toMatchSnapshot();
  });

  it('render with images', () => {
    const images: MetadataImage[] = [{
      src: 'test',
      type: 'test',
      width: 'test',
      height: 'test',
      alt: 'test',
      sizes: 'test',
    }];

    expect(shallow(<App {...defaultProps} images={images}/>)).toMatchSnapshot();
  });

  it('render deep', () => {
    const store = mockStore(DefaultStore);

    const spy = jest.spyOn(App.prototype, 'componentDidMount');

    const component = mount(<Provider store={store}><BrowserRouter><App {...defaultProps} /></BrowserRouter></Provider>);

    expect(spy).toHaveBeenCalled();

    component.unmount();
  });

  it('map state to props', () => {
    const initialState = {
      location: '/',
      modified: data,
    } as StoreTree;

    const mapProps = mapStateToProps(initialState);

    expect(mapProps.location).toEqual(initialState.location);
    expect(mapProps.modified).toEqual(initialState.modified);
  });

  it('map dispatch to props', () => {
    const dispatch = jest.fn();

    mapDispatchToProps(dispatch).update('/');

    expect(dispatch.mock.calls[0][0]).toEqual({type: 'ROUTE_UPDATE', payload: '/'});
  });
});
