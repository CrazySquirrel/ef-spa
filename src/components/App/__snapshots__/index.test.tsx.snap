version https://git-lfs.github.com/spec/v1
oid sha256:f727ee0fad7cef7d00c3573ba93b35ae155a1d6945212403de1aeed9e68d9f84
size 636984
